IPTV Manager integration

CodingBullywug opened the pull request:
### About
[IPTV Manager](https://github.com/add-ons/service.iptv.manager) allows addons such as drnu to integrate live channels and epgs (tv guides) into KODIs PVR ("Live TV") via [Simple IPTV](https://kodi.tv/addons/matrix/pvr.iptvsimple). This pull request facilitates this integration using the existing live channels provided by the drnu addon as well as add new functions/request for fetching the EPG/TVguide (and communication with IPTV Manager).
The integration into Kodi's PVR allows for each swapping between supported channels. This means both channels provided by drnu as well as by other addons.
The integration is seamless in the sense, that drnu can still function without IPTV Manager installed. By default IPTV Manager is disables, and it is only installed, if the user enables it in the drnu settings. The settings also allow the user to select, which of the three supported channels (DR1, DR2 and Ramasjang) should be sent to IPTV Manager.
_See screenshots in comment below._

### Changes
IPTV Manager has a wiki page on [how to integrate existing channels](https://github.com/add-ons/service.iptv.manager/wiki/Integration), which has been followed in this pull request.
The integration provides IPTV Manager with links for live channels and EPG (tv-guide) via sockets. For this purpose IPTV Manager has provided a template class/file (`iptvmanager.py`), which handles the communication. For this purpose, the pull requests adds 4 new functions to addon.py and 2 new routes:
`getIptvLiveChannels` and `getIptvEpg` - get channels and epg from `tvapi.py` and format them correctly for IPTV Manager.
`iptv_channels` and `iptv_epg` - send channels and epg (tvguide) to IPTV Manager using the `iptvmanager` class.
The two routes `plugin://plugin.video.drnu?iptv=channels` and `plugin://plugin.video.drnu?iptv=epg` are used by IPTV Manager to fetch the live channels and epg, respectively.
To get the EPG, two new functions have been added to `tvapi.py`:
`getTvguideChannels` - gets channel IDs needed for requesting EPG from DR.
`getTvguide` - gets the channel TVguides for DRs channels. It is currently hardcoded to get the TVguide from the last 12 hours and the next 12 hours (24 hours in total).
The `settings.xml` now features a new category/tab for IPTV Integration. By default the IPTV Integration is disabled (and not installed). If enabled, it will request the user to install IPTV Manager. The settings also allows the user to select, which DR channels to integrate into Kodi PVR via IPTV Manager.

I have tested the implementation over the last month without any problems. I have tested it on both a Kodi (Matrix) installation on Windows 7 (coding/debugging) and Libreelec on a Raspberry Pi 4 (deployment). I have also tested it along side [the Pluto.tv addon](https://kodi.tv/addons/matrix/plugin.video.plutotv), which also supports IPTV Manager.

### IPTV Manager
IPTV Manager github repo: https://github.com/add-ons/service.iptv.manager
IPTV Manager on Kodi.tv: https://kodi.tv/addons/matrix/service.iptv.manager
IPTV Manager supported addons: https://github.com/add-ons/service.iptv.manager/wiki/Supported-Addons

### Suggestions for further improvement
- Allow users to select lead and lag time for TV Guide in settings.
- Use language files (strings.po) for labels in settings. (What range should be used? 30800-30900 seem unused for now)
- Setting the "Starting area" to `Area selector` seems to cause problems with (at least) the EPG. Exact cause unknown, but I suspect it is the custom widget used, which is opened when the addon starts (with this setting). Possible solution could be adding a less fancy "Area selector" as a fallback for users using IPTV.
- Write a guide on drnu wiki on how to setup it up. (IPTV Manager needs a bit little setup)

### Closing remarks
I hope that you can see the potential in the integration with IPTV Manager, and I will be happy to help continuously support the integration. I can see, that you have made some commits since I forked, and due to the changes, github cannot automatically merge the changes. I will of course be happy to help with sorting out the merge.
I am looking forward to hear your feedback.

CodingBullywug commented:
### Screenshots
Integration into channel list:
![screenshot00003](https://user-images.githubusercontent.com/63240593/173683279-5d8f781e-f5c7-4874-a301-ec05b24f2c28.png)

TV-guide:
![screenshot00001](https://user-images.githubusercontent.com/63240593/173683352-1bd05a91-eaf5-4034-85a1-18aa96362823.png)

Settings:
![screenshot00000](https://user-images.githubusercontent.com/63240593/173683180-d8f1097c-2422-447b-bee0-1b1ce468c266.png)

TermeHansen commented:
Hi @CodingBullywug, Thank you very much for this nice contribution. I don't see any problem in adding this feature to the addon as long as it is done as you suggest via a setting that is default off. 

For the PVr part I do something similar myself, but I use the tvheadend server to serve the channels and get the epg information for kodi PVR - this means that I can also record shows. Anyhow that nothing to do with adding this approach also.

I will go through your PR soon and add comments. I already now see a couple of things I would like to be different - but not a lot, well done!

In regards of the problems with the Starting Area, I have a similar issue when trying to use the addon from yatse app on my mobile. I will use this moment to think of a better way to use our addon from other addons/apps and we can hopefully fix that here also then...

TermeHansen commented on resources/lib/addon.py:
```diff
@@ -553,3 +633,14 @@ def route(self, query):
                 fh.write(traceback.format_exc())
             heading = 'drnu addon crash'
             xbmcgui.Dialog().ok(heading, '\n'.join([tr(30906), tr(30907), tr(30908), crash_file]))
```
to fix the merge I think you just need to change 
`xbmcgui.Dialog().ok(heading, '\n'.join([tr(30906), tr(30907), tr(30908), crash_file]))
`
`to xbmcgui.Dialog().ok(heading, '\n'.join([tr(30906), tr(30907), tr(30908), str(crash_file)]))`

CodingBullywug commented:
@TermeHansen Thank you! I am glad that you are positive towards the pull request. I am looking forward to your comments.

I have resolved the merge conflict (6018f54). Git was having difficulties with both branches changing code at the end of `addon.py`.

Regarding the Area Selector, I think a simple solution would be to create a "simple" Area Selector using `ListItem` and `addDirectoryItems`, similar to how [the main menu](https://github.com/xbmc-danish-addons/plugin.video.drnu/blob/25abe9bb437900232ab6937d8aec7943844b5e2f/resources/lib/addon.py#L124) is created. In the settings, we could have a button toggle using the Fancy (Current) or Simple (ListItem) Area Selector. Default could be the Fancy area selector to avoid breaking anything for existing users. The tricky part might be to disentangle the current Area Selector from the addon. I don't mind giving it a shot. But I think, I would be better to create a separate pull request for it.

TermeHansen commented:
@CodingBullywug i have tried to ping you in a comment in
https://github.com/xbmc-danish-addons/plugin.video.drnu/pull/68

But I'm not sure you get notified by it. In tge new api I have updated function to get livestream links (with _ull -> ultra low latency) and also added a schedule pull function. I also added a timeout to first areachooser, so it should work with your usecase, it works for the android kodi remote now at least :)

Could you please have a look and give suggestions to streamline for your need and also just test the current state?

CodingBullywug commented:
@TermeHansen Sorry for not getting back to you sooner, I was manually checking this PR for updates, so I didn't notice your tag in the other PR.

I have skimmed your changes in #68. There were no immediate issues, that caught my eye, but I'll have a closer look soon. Nice that you have included a schedule function as well :-)
I will also try to re-work the IPTV Manager integration to include the new api.

Regarding the Area Selector, I have been using the solution I mentioned above since June. See c5cf04d, branch [areaselector-simple](https://github.com/CodingBullywug/plugin.video.drnu/tree/areaselector-simple) on my fork. I recently discovered works great with Yatse as well (fantastic app btw, was using Kore before, but found it a little lackluster). The solution implements a "simple" area selector, which just uses the `contextMenuItem` to add the four areas. See image at the end of this comment. It adds four new routes - one for each area. I found this convenient for linking to a specific area, when using the skin `arctic zephyr reloaded`. The solution also adds a setting for selecting between the two area selectors (default being the old "fancy" area selector).
Let me know, what you think of this solution compared to the timeout. Maybe we should create a new PR to discuss it separately :-)

Unfortunately, I am a little busy at the moment, but I will have a look at the new API as soon as possible, but it will most likely not be until next week (earliest).

![image](https://user-images.githubusercontent.com/63240593/189203069-be24c9b1-f189-4851-8c2a-7fc2461389b8.png)

TermeHansen commented:
After using it for more than a week now on production (family tv) with no real issues, I decided last night to push it as version 6.0.0 in the official repo. So soon out as a normal update..

But please let me know if you find something and we can make a fast bug release after.

I will look at your suggestion on simple menu and probably implement it

TermeHansen commented:
@CodingBullywug just FYI I just updated to version 6.0.1 with some small fixes and addind your simpleareashooser, and fixed the schedule function to also work for up to 7 day forward...

https://github.com/xbmc-danish-addons/plugin.video.drnu/pull/70/files

CodingBullywug commented:
@TermeHansen Alright, I am 99% done with updating the PR to the new API, but a have a few questions (in **bold** below) and a little "long term" testing to do before, I am confident in merging it with the main branch. I will be "long term" testing it on production (also family TV) for the next week. If no problems arise, I'll be confident in it being merged.

Your implementation of the new API made it very easy to adapt the old code in `addon.py` to fit the new API. The new API has actually simplified the IPTV related code in `addon.py`. I also appreciate, that you have included DRTV and DRTV Ekstra. I had thought about DRTV myself, but have not had time to look into it. Very cool, that you have included those as well.
I noticed, that the channels returned by getLiveChannels all seem to have the same "id" in the root of their structure. **Is this intentional or a bug?** I suspect, that it might be a user id, as each channel has a unique id further into the dictionary structure.

For some of the programs in the schedule, there is a `watchPath` and `path` with relative links to the programs can be viewed on demand. IPTV Manager supports linking to these, but [it requires a direct link to the stream](https://github.com/add-ons/service.iptv.manager/wiki/JSON-EPG-format). **Do you know if the paths can be (easily) resolved to streams?**

I only changed a minor thing in `tvapi.py` - adding the `duration` argument to the schedule, as you quickly noticed and improved on by adding support for up to 7 days :-) I have added it as a setting for the user in the addon settings.
I noticed, that when using more than 1 day, there are multiple entries for each channel, as it fetches one day at a time. A bit odd at first, but it was easy to handle (See [c5f95b0](https://github.com/xbmc-danish-addons/plugin.video.drnu/pull/67/commits/c5f95b017d16b2001cf81ba41c9f5b127a0d574c#diff-c025d50cfc229559497d304d335832578dfec5f228373aba5eb1764d62d127e9R293-R297)). Intuitively, I would only expect one entry per channel, but as long as it is only IPTV Manager, which uses the function (with more than 24 hours), it does not really matter, where the entries are merged.

I have also added language support to the settings. I selected the range 30550 to 30563.  The other settings seemed to end at 30534, and I wanted to leave some room for additional strings related to the other settings. Let me know, if I should use another range of numbers.
I had an issue with string 30563, which I wanted to use for a default value of a "Refresh interval" setting. The setting basically tells the user, that it needs to be set in IPTV Manager. I think it is useful information to have in the settings, but as far as I can tell, default values does not accept string ids. The new addon settings structure might support it. For now I have left it in English, and "reserved" 30563 for later use.

TermeHansen commented:
for your questions:

I made the update for the schedule function because I use it myself now to generate an xmltv file for tvheadend, and for that it doesn't matter how programs from different channels are ordered, so I didn't bother to make the function more complex just to keep them in each channels list - but yes I also noticed and thought let's fix if needed :D 


I don't know what that root id is in the returned entry from the items from the programcard call. I don't think it is related to the deviceid that we generate for each client, as that is not used in the programcard request - anyhow, not a bug and not important IMHO.

For the part with the stream url (if exists) it is very easy with this api from dr. Just use the `api.get_stream(id)` with the id of the item:

```
{'schedules': [{'id': '2ed0e06',
    'channelId': '20875',
    'startDate': '2022-09-22T19:55:00Z',
    'endDate': '2022-09-22T21:25:00Z',
    'startTimeInDefaultTimeZone': '2022-09-22T21:55:00+02:00',
    'endTimeInDefaultTimeZone': '2022-09-22T23:25:00+02:00',
    'item': {'id': '337880',
     'watchPath': '/se/vera_-det-ligger-i-blodet_337880',
     'path': '/episode/vera_-det-ligger-i-blodet_337880',
     'title': 'Vera: Det ligger i blodet', 
```

```
> api.get_stream(337880)

{'accessService': 'StandardVideo',
 'channels': 2,
 'complianceProfile': 'drtv-avc-high, 41',
 'containerFormat': 'FMP4',
 'deliveryType': 'Stream',
 'drm': 'PlayReady',
 'format': 'video/hls',
 'resolution': 'HD-1080',
 'height': 0,
 'isStreamLive': False,
 'language': 'foreign',
 'name': 'Video Resource',
 'subtitles': [{'format': 'text/vtt;charset=utf-8',
   'isEmbedded': True,
   'isHardSubs': False,
   'language': 'ForeignLanguageSubtitles',
   'link': 'https://drod24m.akamaized.net/all/clear/none/b3/6323f928aa5a6103f8b12bb3/00021961690/subtitles/Foreign-17363635-dfc4527f-de5c-49d1-a530-980742faa7b6.vtt'}],
 'timeCodes': [],
 'url': 'https://drod24m.akamaized.net/dk/encrypted/none/b3/6323f928aa5a6103f8b12bb3/00021961690/stream_fmp4/master_manifest.m3u8',
 'width': 0}
```


I'll have a look over the weekend on the PR as it is now, and might come the some re-factor suggestions...

TermeHansen commented:
Ok, I just activated the github CI for this PR, as it was not running because it was your first PR here :)
As you can see the official kodi addon checker failed with something wrong in the language po files in line 213 - not sure what, will you check?

I also cloned your repo and branch to test and re-factor a bit. I am not allowed to push to your fork, so here is a patch file that I would like you to consider to take in to your branch.
[code-refactor.patch.zip](https://github.com/xbmc-danish-addons/plugin.video.drnu/files/9638878/code-refactor.patch.zip)

CodingBullywug commented:
Thank you for the re-factored code. I've added the changes manually as I am not familiar with working with patch'es. But I completely agree with all your suggestions. I makes the code more streamlined. Thank you for taking your time to review it.

I have added the stream to the IPTV epg (e1e3054), and it works very nicely with IPTV Manager. Rather than calling `api.get_stream(...)` for each item in the schedule, I have used a url to the plugin (`?playVideo=...`) instead. Then the api will only make a session call to resolve the id, when it is needed - and not for each item in the schedule, when it is updated. I simply copied the relevant code part from the `kodi_item()` function.
Unfortunately, the line becomes too long (142>130) to pass the CI (109e00a). I have two ideas for solving it.
1) Set `item = schedule['item']`, and use that for building the `stream_path`. This can also replace `schedule['item']` other places in the same code branch.
2) Keep using `schedule['item']`, but build `stream_path` over multiple rows. I.e.:
```
stream_path = self._plugin_url
stream_path += f"?playVideo={schedule['item']['id']}&kids={str(kids)}&idpath={schedule['item']['path']}"
```
I think that 1) might fit the coding style best - especially after seeing your re-factoring suggestions. If you have a third option, I am also open for that.

I agree, that the root id in the program card is not important. I was just curious :-)

Regarding the multiple entries for the same channel in `api.get_schedules()`, I think your suggestion in the code re-factor solves it elegantly in the `addon.py`. No need to change `api.get_schedules()`.

TermeHansen commented:
@CodingBullywug I made a code suggestion, I hope you can see?

CodingBullywug commented:
@TermeHansen I don't think, I can see the code suggestion. It's not showing up here, and I'm not sure where else to look :-)
Can you throw me link?

I have noticed a problem with the get_schedule(). The hour argument is subtracted by 2, before the request is being sent to DR's api. This means that for midnight and 1am, the hour requested becomes negative, which causes an error message from DR.
I think we need to do the subtraction using the datetime library and before separating the date and hour. It can become a little convoluted, when we have both date and hour as input arguments.
I have not looked into a solution yet - I think it also needs to fit with your setup.

TermeHansen commented:
You are very right about that bug in the substraction of 2 hours, that I decided to do, to be sure to get epg also of shows running now, so by this get shows that have started from 2 hours before now. Anyhow it is true with this bug as it will crash if hour is less than 2 - brainfart...

```
+++ b/resources/lib/tvapi.py
@@ -323,7 +323,7 @@ class Api():
 
     def get_schedules(self, channels=CHANNEL_IDS, date=None, hour=None, duration=6):
         url = URL + '/schedules?'
-        now = datetime.now()
+        now = datetime.now() - timedelta(hours=2) # be sure to get epg also for shows start within 2 hours from now
         if date is None:
             date = now.strftime("%Y-%m-%d")
         if hour is None:
@@ -332,7 +332,7 @@ class Api():
         if duration <= 24:
             data = {
                 'date': date,
-                'hour': hour-2,
+                'hour': hour,
```

the other part I have given as a code suggestion if you look under files changed in the top of this PR and look for my suggestion under the addon.py changes. Let me know if you still cannot find it, then I will share the code here as well :)

CodingBullywug commented on resources/lib/addon.py:
```diff
@@ -226,6 +225,52 @@ def showRecentlyWatched(self):
+                    schedule_dict['stream'] = stream_path
+                channel_epg.append(schedule_dict)
+            epg[channel_epg_id] += channel_epg
+        return epg
```
test comment

CodingBullywug commented:
> the other part I have given as a code suggestion if you look under files changed in the top of this PR
> https://github.com/xbmc-danish-addons/plugin.video.drnu/pull/67/files
> and look for my suggestion under the addon.py changes. Let me know if you still cannot find it, then I will share the code here as well :)

I was not aware of that feature on Github. Thank you for letting me know.
But I have looked and looked, and I can not find your comment/suggestion. I have even added a test comment on my own. It is very visible, so I am pretty sure, that I have not overlooked you comment ;-)
Please share the code here :-)

TermeHansen commented:
yes, indeed that is weird, but also that I did not see it here, made me think that it was not working as anticipated - maybe todo with that I am not the repo owner but only the admin... Anyhow

``` 
               if ('path' in schedule['item']):
                    schedule_dict['stream'] = "{}?playVideo={}&kids={}&idpath={}".format(
                        self._plugin_url,
                        schedule['item']['id'],
                        self.api.kids_item(schedule['item']),
                        schedule['item']['path'],
                    )
```

Here is how it looks from my side :smile: 
![image](https://user-images.githubusercontent.com/6922018/192161441-a6467ca3-7623-4344-9900-ffd6bb9e852f.png)

TermeHansen commented:
Alright it's starting to come together I think. Let me know when you have used it a bit in production, and then we can merge it in main...

CodingBullywug commented:
Awesome! I've added it, tested it and pushed it. And it seems to pass all checks! Yay! :-)

I still don't understand, why your suggestion it is not visible to me/all :D But it seems out of our hands. I have checked the "Allow edits by maintainers" checkbox to see if that made a difference, but it does not seem to matter. Anyway... Not overly important :-)

I have also changed the visibility of the IPTV settings. Before, all the settings were hidden until IPTV Manager were installed. That felt weird. Now, they are visible, but disabled until IPTV Manager is installed. That should give new users an idea about what it is about without having to install IPTV Manager. I have also changed the default to disabled for IPTV Manager integration even if IPTV Manager is already installed.
Overall, I think, the changes to the settings are more honest/genuine towards the users of the addon.

CodingBullywug commented:
> Alright it's starting to come together I think. Let me know when you have used it a bit in production, and then we can merge it in main...

Sounds great! I'll test it the next week, so if nothing happens, maybe next weekend we can merge it?

TermeHansen commented:
@CodingBullywug Is this ready to merge into main?

CodingBullywug commented:
> @CodingBullywug Is this ready to merge into main?

@TermeHansen I consider it ready to merge now.
I have tested the changes in 008b0ce over the last week and a half. The previous error is gone and have not run into any new problems. I have fetched the schedule pretty aggressively every hour and fetched 72 hours each time.
